docs/design/document-type.md

```diff
@@ -14,346 +14,164 @@ such as `struct`, `union`, etc. which are handled by the [Kotlin Smithy SDK](kot
-
-    /**
-     * Returns [index] th element of an array as [SmithyArray] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getArray(index: Int) = content[index] as? SmithyArray
-
-    /**
-     * Returns [index] th element of an array as [SmithyMap] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getMap(index: Int) = content[index] as? SmithyMap
-
-    override fun toString(): String = content.joinToString( separator = ",", prefix = "[", postfix = "]" )
-}
-
-/**
- * Class representing document `map` type
- *
- * Map consists of name-value pairs, where the value is an arbitrary Document. This is much like a JSON object.
- */
-data class SmithyMap(val content: Map<String, Document>): Document(), Map<String, Document> by content {
-    /**
-     * Returns [SmithyNumber] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getNumber(key: String): SmithyNumber? = getValue(key) as? SmithyNumber
-
-    /**
-     * Returns [SmithyBool] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getBoolean(key: String): SmithyBool? = getValue(key) as? SmithyBool
-
-    /**
-     * Returns [SmithyString] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getString(key: String): SmithyString? = getValue(key) as? SmithyString
-
-    /**
-     * Returns [SmithyArray] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getArray(key: String): SmithyArray? = getValue(key) as? SmithyArray
-
-    /**
-     * Returns [SmithyMap] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getMap(key: String): SmithyMap? = getValue(key) as? SmithyMap
-
-    override fun toString(): String {
-        return content.entries.joinToString(
-            separator = ",",
-            prefix = "{",
-            postfix = "}",
-            transform = {(k, v) -> """"$k":$v"""}
-        )
-    }
-}
-
-fun Boolean.toDocument() = SmithyBool(this)
-fun Number.toDocument() = SmithyNumber(this)
-fun String.toDocument() = SmithyString(this)
-
-/**
- * DSL builder for a [SmithyArray]
- */
-class DocumentArrayBuilder internal constructor() {
-    internal val content: MutableList<Document> = mutableListOf()
-
-    /**
-     * Adds [this] value to the current [SmithyArray] as [SmithyString]
-     */
-    operator fun String.unaryPlus() {
-        content.add(SmithyString(this))
+    companion object {
+        /**
+         * Overloads to create a document from possible values.
+         */
+        operator fun invoke(init: Builder.() -> Unit): Document
+        operator fun invoke(value: kotlin.Number): Document
```

Comment:
these extra overloads for types with an explicit variant seem unnecessary to me.

```kotlin
val x = Document(12)
```

vs explicit

```kotlin
val x = Document.Number(12)
```

If we felt strongly about keeping them I'd remove these as `operator invoke` overloads to free "builder" functions outside of the class definition like:

```kotlin
sealed class Document { ... }

/**
 * Construct a document from a [Number]
 */
fun Document(value: kotlin.Number): Document = Document.Number(value)
...
```

In part because we should move anything unrelated to the core type outside of the type definition where possible but mostly because intellij will still suggest these functions to you but then insert things like `Document.invoke{...}` instead of what you want `Document(...)`